resp/resp.go

```diff
@@ -204,12 +214,26 @@ func readAllAppend(r io.Reader, b []byte) ([]byte, error) {
 	// bytes.MinRead) and the data fits into b.
 	b = append(b, buf.Bytes()...)
 	putBuffer(buf)
-	if b == nil {
-		b = []byte{}
-	}
 	return b, err
 }
 
+// readNAppend appends exactly n bytes from r into b.
+func readNAppend(r io.Reader, b []byte, n int) ([]byte, error) {
+	// ReadFull does not call Read if n == 0, but some readers (like
+	// limitedReaderPlus) need at least one call to Read, so we just
+	// call Read with an empty []byte if n == 0.
+	if n == 0 {
+		_, err := r.Read(b[len(b):])
+		return b, err
+	} else if m := cap(b) - len(b); n > m {
+		// Go 1.11 optimizes append(x, make([]T, y)...) calls by
+		// avoiding unnecessary allocations and copies. (CL 109517)
+		b = append(b[:cap(b)], make([]byte, n-m)...)[:len(b)]
```

Comment:
In this file there's already a function called `expand`, which is used other places. I _think_ this logic could be moved into there, and that function could then be used here too, like:

```
b = expand(b, len(b)+n)
_, err := io.ReadFull(r, b[len(b):len(b)+n])
return b, err
```